### DISCUSSION TITLE
What limits are placed on names in `$defs` and data type names?

### DISCUSSION THREAD

**Discussion by sjlongland** [json-schema-org/community#185] [Q&A] [answered]
Hi all,

This is a bit of a silly question, but a cursory read of the draft RFC for JSONSchema and various documentation items has thus far not answered it for me.

What names are valid in the name of a `$defs` member or the name used in a field data type?

My intended use case is to basically be able to describe a SQL database tables through JSONSchema documents.  The goal being a kind-of HTTP-based ORM, which coupled with a suitably specified ReST API specification and using [JSONSchema-based form generators](http://json-schema.org/implementations.html#web-ui-generation) for the front-end, would pretty much give you a pretty powerful declarative ORM web-application framework without having to write much in the way of SQL or JavaScript/HTML code.

Specifically I'd like to be able to encode both "generic" types (e.g. `string`, `number`, `integer`…) and more specific types; e.g. maybe `postgres:ltree` for a PostgreSQL [ltree](https://www.postgresql.org/docs/current/ltree.html) column -- so that basically the type name is "scoped", and won't clash with anyone else's data types.

Are there any gotchas I need to be aware of?

**Comment by handrews** [comment, marked as answer]
Interesting!  I'm doing something similar (I think) with Apache Tinkerpop's Gremlin query language for graph databases.

You might want to bring this discussion to the [Slack](https://json-schema.org/slack) discussion.

`$defs` names are just JSON object names and can be anything that is syntactically legal in that sense.  Of course, some names will require URI-encoding to be used in JSON Pointer fragments, but if you're using a proper implementation of JSON Pointer and constructing URIs with a proper URI library, that won't be an issue.

I'm not sure what you mean by "name used in a field data type"?  The `type` keyword is **not** extensible.

**Reply by sjlongland** [reply]
Ahh right, so for `$defs` names, if I stick to what can be sanely URI-encoded I should be fine.

And understood regarding `type`, can I put other annotations alongside the `$defs` to further refine the data type?  e.g. if I do

```json
{
  "$id": "https://example.com/blog-category.json",
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "description": "An example of a blogging category",
  "type": "object",
  "properties": {
    "category-id": {
      "$ref": "#/$defs/postgres-uuid"
    },
    "category-path": {
      "$ref": "#/$defs/postgres-ltree"
    },
    "name": {
      "type": "string"
    }
  },
  "$defs": {
     "postgres-uuid": {
         "type": "string",
         "columnType": "uuid",
         "displayFormat": "base64"
     },
     "postgres-ltree": {
         "type": "string",
         "columnType": "ltree"
     }
  }
}
```

… that `columnType` and `displayFormat` value will be associated with the `category-id` and `category-path` fields defined in the object?
